File: sdks/typescript/test/wire-fixtures.test.ts
/**
 * Every shared fixture, through every parser that claims to produce its type.
 *
 * See `sdks/fixtures/README.md`. The short version: check-parity.py compares
 * the three SDKs' *declarations* and can only see field names, which is how
 * `approval` stayed wrong in two of them while the check reported ok. These
 * compare values.
 *
 * And they run each fixture through *both* parsers where there are two.
 * `Session` is parsed once in client.ts for REST and once in stomp.ts for the
 * WebSocket, and the two are byte-identical copies with nothing holding them
 * that way. Python had the same pair and they had already drifted: the
 * WebSocket one returned raw strings where REST returned instants.
 */

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, readdirSync } from "node:fs";
import { join } from "node:path";
import { fileURLToPath } from "node:url";

import {
  parseAccount, parseConnection, parseHolding, parseMarket, parseMarketplace,
  parseOrder, parsePerson, parseSecurity, parseSession, parseToken,
  embeddedOrders,
} from "../src/client.ts";
import { parseSession as parseSessionOverWs } from "../src/stomp.ts";

type Json = Record<string, unknown>;

const FIXTURE_DIR = join(fileURLToPath(new URL(".", import.meta.url)), "..", "..", "fixtures");

/** type -> the parsers that must all agree about it, named so a failure says which. */
const PARSERS: Record<string, [string, (data: Json) => unknown][]> = {
  Order: [["client", (d) => parseOrder(d)]],
  // Not a parser but a shape the SDK has to recognise, which is why it has
  // broken twice -- and silently here both times, since `._embedded` on an
  // array is undefined rather than an error.
  OrdersSnapshot: [["client", (d) => ({ orders: embeddedOrders(d).map((o) => parseOrder(o)) })]],
  Session: [["client", (d) => parseSession(d)], ["stomp", (d) => parseSessionOverWs(d)]],
  Holding: [["client", (d) => parseHolding(d)]],
  Account: [["client", (d) => parseAccount(d)]],
  Person: [["client", (d) => parsePerson(d)]],
  Market: [["client", (d) => parseMarket(d)]],
  Marketplace: [["client", (d) => parseMarketplace(d)]],
  ClientConnection: [["client", (d) => parseConnection(d)]],
  Security: [["client", (d) => parseSecurity(d)]],
  Token: [["client", (d) => parseToken(d)]],
};

interface Fixture {
  name: string;
  type: string;
  why: string;
  payload: Json;
  expect: Json;
}

const FIXTURES: Fixture[] = readdirSync(FIXTURE_DIR)
  .filter((f) => f.endsWith(".json"))
  .sort()
  .map((f) => ({
    name: f.replace(/\.json$/, ""),
    ...(JSON.parse(readFileSync(join(FIXTURE_DIR, f), "utf8")) as Omit<Fixture, "name">),
  }));

function isInstant(v: unknown): v is { epochMilli: number } {
  return typeof v === "object" && v !== null && !Array.isArray(v)
    && Object.keys(v).length === 1 && "epochMilli" in v;
}

function field(parsed: unknown, wireName: string): unknown {
  assert.ok(parsed !== null && typeof parsed === "object", `expected an object, got ${parsed}`);
  assert.ok(wireName in (parsed as Json), `no field '${wireName}' on ${JSON.stringify(parsed)}`);
  return (parsed as Json)[wireName];
}

function check(parsed: unknown, expected: unknown, where: string): void {
  if (isInstant(expected)) {
    assert.ok(parsed !== null && parsed !== undefined, `${where}: expected an instant, got ${parsed}`);
    assert.ok(
      parsed instanceof Date,
      `${where}: expected a Date, got ${typeof parsed} ${JSON.stringify(parsed)} `
        + `-- the parser left the wire value unconverted`,
    );
    assert.ok(!Number.isNaN(parsed.getTime()), `${where}: Invalid Date`);
    assert.equal(
      parsed.getTime(), expected.epochMilli,
      `${where}: ${parsed.toISOString()} is ${parsed.getTime()}, expected ${expected.epochMilli} `
        + `(${new Date(expected.epochMilli).toISOString()}). A bare server timestamp means UTC; `
        + `new Date(bare) reads it as local and is right only in UTC.`,
    );
    return;
  }

  if (Array.isArray(expected)) {
    assert.ok(Array.isArray(parsed), `${where}: expected a list, got ${JSON.stringify(parsed)}`);
    assert.equal(parsed.length, expected.length, `${where}: wrong length`);
    expected.forEach((want, i) => {
      if (want !== null && typeof want === "object" && !Array.isArray(want)) {
        for (const [key, value] of Object.entries(want as Json)) {
          check(field(parsed[i], key), value, `${where}[${i}].${key}`);
        }
      } else {
        check(parsed[i], want, `${where}[${i}]`);
      }
    });
    return;
  }

  if (expected !== null && typeof expected === "object") {
    for (const [key, value] of Object.entries(expected as Json)) {
      check(field(parsed, key), value, `${where}.${key}`);
    }
    return;
  }

  // null must stay null: `?? false` and `?? 0` turn "undecided" into a decision.
  assert.equal(parsed, expected, `${where}: expected ${JSON.stringify(expected)}, got ${JSON.stringify(parsed)}`);
}

for (const fixture of FIXTURES) {
  for (const [parserName, parse] of PARSERS[fixture.type] ?? []) {
    test(`${fixture.name} / ${parserName}`, () => {
      const parsed = parse(fixture.payload);
      assert.ok(parsed !== null && parsed !== undefined, `${parserName} returned ${parsed}`);
      for (const [key, expected] of Object.entries(fixture.expect)) {
        check(field(parsed, key), expected, `${fixture.name}/${parserName}.${key}`);
      }
    });
  }
}

test("there are fixtures to run", () => {
  // Guard the guard: a bad path would report everything passing.
  assert.ok(FIXTURES.length >= 10, `only found ${FIXTURES.length} fixtures`);
});

test("every fixture type has a parser", () => {
  // A fixture for a type nothing parses is a test that does not run.
  const unmapped = [...new Set(FIXTURES.map((f) => f.type))].filter((t) => !(t in PARSERS));
  assert.deepEqual(unmapped, [], `fixtures exist for types no parser here handles`);
});

File: sdks/java/fm-sdk/src/test/java/fm/internal/WireFixturesTest.java
package fm.internal;

import fm.model.Account;
import fm.model.ClientConnection;
import fm.model.Holding;
import fm.model.Market;
import fm.model.Marketplace;
import fm.model.Order;
import fm.model.Person;
import fm.model.Security;
import fm.model.Session;
import fm.model.Token;
import java.io.IOException;
import java.lang.reflect.Method;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import tools.jackson.databind.JsonNode;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

/**
 * Every shared fixture, read with the mapper the SDK actually uses.
 *
 * <p>See {@code sdks/fixtures/README.md}. The short version: check-parity.py
 * compares the three SDKs' declarations and can only see field names, which is
 * how {@code approval} stayed wrong in Python and TypeScript while the check
 * reported ok. These compare values.
 *
 * <p>Java has one parser per type — Jackson, through {@link
 * HttpFlexemarkets#MAPPER}, which {@link Events} is handed too — so unlike the
 * other two SDKs there is no REST/WebSocket pair to hold together here. What
 * this does catch is the annotations: every record binds the wire through a
 * {@code @JsonCreator} taking {@code String} timestamps, and every type carries
 * {@code @JsonIgnoreProperties(ignoreUnknown = true)}. That combination fails
 * quietly — a renamed component binds to nothing and arrives as null rather
 * than throwing.
 */
class WireFixturesTest {

    private static final Path FIXTURES =
            Path.of("..", "..", "fixtures").toAbsolutePath().normalize();

    /** Wire type name -> the record the SDK deserializes it into. */
    private static final Map<String, Class<?>> TYPES = Map.of(
            "Order", fm.model.Order.class,
            "Session", fm.model.Session.class,
            "Holding", fm.model.Holding.class,
            "Account", fm.model.Account.class,
            "Person", fm.model.Person.class,
            "Market", fm.model.Market.class,
            "Marketplace", fm.model.Marketplace.class,
            "ClientConnection", fm.model.ClientConnection.class,
            "Security", fm.model.Security.class,
            "Token", fm.model.Token.class);

    /**
     * The snapshot envelope is not a record the mapper binds -- it is a shape
     * the SDK has to recognise, which is exactly why it has broken twice. The
     * fixtures for it run through the real unwrap.
     */
    record OrdersSnapshot(java.util.List<fm.model.Order> orders) { }

    /** Types the fixture run handles by a route other than {@link #TYPES}. */
    private static final java.util.Set<String> HANDLED_WITHOUT_A_TYPE =
            java.util.Set.of("OrdersSnapshot");

    record Fixture(String name, String type, JsonNode payload, JsonNode expect) {
        @Override
        public String toString() {
            return name + " -> " + type;
        }
    }

    static Stream<Fixture> fixtures() throws IOException {
        try (var files = Files.list(FIXTURES)) {
            var out = new ArrayList<Fixture>();
            for (var path : files.filter(p -> p.toString().endsWith(".json")).sorted().toList()) {
                var doc = HttpFlexemarkets.MAPPER.readTree(Files.readString(path));
                var name = path.getFileName().toString().replace(".json", "");
                out.add(new Fixture(name, doc.get("type").asString(),
                                    doc.get("payload"), doc.get("expect")));
            }
            return out.stream();
        }
    }

    @ParameterizedTest(name = "{0}")
    @MethodSource("fixtures")
    void fixture(Fixture fixture) {
        Object parsed;
        if ("OrdersSnapshot".equals(fixture.type())) {
            parsed = new OrdersSnapshot(HttpFlexemarkets._unwrapOrders(
                    new fm.Snapshot<>(fixture.payload(), fm.Snapshot.NO_SEQ)).body());
        } else {
            var type = TYPES.get(fixture.type());
            assertNotNull(type, "no Java type mapped for " + fixture.type());
            parsed = HttpFlexemarkets.MAPPER.treeToValue(fixture.payload(), type);
        }
        assertNotNull(parsed, fixture.name() + ": deserialized to null");

        fixture.expect().propertyStream().forEach(entry ->
                _check(_read(parsed, entry.getKey()), entry.getValue(),
                      fixture.name() + "." + entry.getKey()));
    }

    /** The value of one record component, by its wire name. */
    private static Object _read(Object parsed, String name) {
        Method accessor;
        try {
            accessor = parsed.getClass().getMethod(name);
        } catch (NoSuchMethodException e) {
            return fail(parsed.getClass().getSimpleName() + " has no component '" + name + "'");
        }
        try {
            return accessor.invoke(parsed);
        } catch (ReflectiveOperationException e) {
            return fail("could not read " + name, e);
        }
    }

    private static void _check(Object actual, JsonNode expected, String where) {
        if (expected.isObject() && expected.has("epochMilli") && expected.size() == 1) {
            assertNotNull(actual, where + ": expected an instant, got null");
            var instant = assertInstanceOf(Instant.class, actual,
                    where + ": expected an Instant, got " + actual.getClass().getSimpleName()
                            + " " + actual + " -- the wire value was left unconverted");
            assertEquals(expected.get("epochMilli").asLong(), instant.toEpochMilli(),
                    where + ": " + instant + " is not "
                            + Instant.ofEpochMilli(expected.get("epochMilli").asLong()));
            return;
        }

        if (expected.isArray()) {
            assertNotNull(actual, where + ": expected a list, got null");
            // Java spells some of these as arrays (Person.roles) and some as
            // List. The fixture does not care which, only what is in it.
            var list = actual instanceof Object[] array
                    ? List.of(array)
                    : assertInstanceOf(List.class, actual, where + ": expected a List or array");
            assertEquals(expected.size(), list.size(), where + ": wrong length");
            for (int i = 0; i < expected.size(); i++) {
                var want = expected.get(i);
                if (want.isObject()) {
                    int index = i;
                    want.propertyStream().forEach(e ->
                            _check(_read(list.get(index), e.getKey()), e.getValue(),
                                  where + "[" + index + "]." + e.getKey()));
                } else {
                    _check(list.get(i), want, where + "[" + i + "]");
                }
            }
            return;
        }

        if (expected.isObject()) {
            expected.propertyStream().forEach(e ->
                    _check(_read(actual, e.getKey()), e.getValue(), where + "." + e.getKey()));
            return;
        }

        if (expected.isNull()) {
            // Null must survive as null. A primitive component cannot hold it,
            // which is the point: `approval` had to be boxed for exactly this.
            assertEquals(null, actual, where + ": expected null, got " + actual);
            return;
        }
        if (expected.isBoolean()) {
            assertEquals(expected.asBoolean(), actual, where);
            return;
        }
        if (expected.isNumber()) {
            assertNotNull(actual, where + ": expected " + expected.asLong() + ", got null");
            assertEquals(expected.asLong(), ((Number) actual).longValue(), where);
            return;
        }
        // Enums compare by name, so a parser that never converted is visible.
        assertEquals(expected.asString(), String.valueOf(actual), where);
    }

    @Test
    @DisplayName("there are fixtures to run")
    void thereAreFixturesToRun() throws IOException {
        // Guard the guard: a bad path would report everything passing.
        var found = fixtures().toList();
        assertTrue(found.size() >= 10, "only found " + found.size() + " fixtures in " + FIXTURES);
    }

    @Test
    @DisplayName("every fixture type has a Java type")
    void everyFixtureTypeIsMapped() throws IOException {
        // A fixture for a type nothing maps is a test that does not run.
        var unmapped = fixtures().map(Fixture::type).distinct()
                .filter(t -> !TYPES.containsKey(t) && !HANDLED_WITHOUT_A_TYPE.contains(t))
                .sorted().toList();
        assertEquals(List.of(), unmapped, "fixtures exist for types with no Java mapping");
    }
}
